[Wee Jun Hong] iP

whoisjunhong opened the pull request:
# Duke PR from whoisjunhong to nus-cs2103 💯 

- This pull request contains all the changes that has been made on **whoisjunhong/master**
- All commits made are in the commit tab

1. Currently, this iP is a _work in progress_, to see the rest of the project click [here](https://github.com/whoisjunhong/ip)
2. However, comment ~~where you see fit~~, both good and bad practices
3. For example, if you wanna comment on a code like `someFunction()`, do use a fenced code block as shown below to state the issue:

```java
int someFunction( ) {
    return 0;
}
```

# Duke Task List Examples
- [x] Buy stationaries ✏️ 
- [ ] Finish this week's iP task (by: 2022-02-10)
- [ ] Go for lectures (at: 2022-02-05)

> If you wanna go fast, do alone, if you wanna go far, do together. 
> -> Someone important, probably

Commit:
Add Gradle support

whoisjunhong committed:
Level 1

whoisjunhong committed:
Level-2

whoisjunhong committed:
Level-2

whoisjunhong committed:
Level-3

whoisjunhong committed:
ToDos, Events, Deadlines

whoisjunhong committed:
Automated Text UI Testing

whoisjunhong committed:
Level-6: Delete

whoisjunhong committed:
Level-5

whoisjunhong committed:
Use Enums

whoisjunhong committed:
Level-7 Save

whoisjunhong committed:
Level-8

whoisjunhong committed:
Fix merge conflicts

whoisjunhong committed:
A-MoreOOP: Use More OOP

whoisjunhong committed:
A-Packages: Organize into Packages

whoisjunhong committed:
Added JUnit tests for TaskList and Task class

whoisjunhong committed:
Create JAR file

whoisjunhong committed:
Fix text formatting

whoisjunhong committed:
JavaDoc

whoisjunhong committed:
Add JavaDoc

whoisjunhong committed:
Fix Coding Standard

whoisjunhong committed:
Add find function

whoisjunhong committed:
Fix ui

whoisjunhong committed:
Merge branch 'branch-A-Coding-Standard'

whoisjunhong committed:
Merge branches branch-A-JavaDoc, branch-A-Coding-Standard and branch-Level-9

whoisjunhong committed:
Fix commands

junjunjieOng commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,98 @@
+        String command = userInput.split(" ")[0].toLowerCase(Locale.ROOT);
+
+        switch (command) {
+            case "list":
```
Hi 👋 , should there be no indentation for case clause?

junjunjieOng commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public boolean addToDoTask(String userInput) {
+        String description = TaskValidator.validateToDo(userInput);
+        boolean noDescription = description.equals("");
```
Perhaps a variable name sound more like boolean?
```   
boolean hasDescription = description.equals("");
```

junjunjieOng commented on src/main/java/duke/validation/TaskValidator.java:
```diff
@@ -0,0 +1,143 @@
+     * @throws DukeException if Deadline input from user is invalid
+     */
+    public static String[] validateDeadline(String userInput) {
+        String[] information = new String[2];
```
Perhaps a more detailed variable name to understand better?
```
   String[] userInputInformations = new String[2];
```

junjunjieOng commented on src/main/java/duke/validation/TaskValidator.java:
```diff
@@ -0,0 +1,143 @@
+     * @throws DukeException if date input from user is invalid
+     */
+    public static LocalDate convertDate(final String date) throws DukeException {
+        boolean valid = false;
```
Should this valid = false be deleted?  It seems that it is not used in this method.😅

junjunjieOng reviewed (commented):
Overall code is neat and clear to understand. Just some minor fix to match the coding standard.

e0544333 commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,41 @@
+                return;
+            }
+
+            System.out.println("Here are the matching tasks in your list containing "
+                                + "\"" + keyword + "\"" + ":");
```
Like the unique response from the bot!
I was wondering if you could shift this print statement to another class?
A suggestion would be the ui.java class.

Otherwise, excellent work!

joshuayapwj98 replied:
Like the unique response from the bot!
I was wondering if you could shift this print statement to another class?
A suggestion would be the ui.java class.

Otherwise, excellent work!

joshuayapwj98 commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,41 @@
+                                + "\"" + keyword + "\"" + ":");
+
+            Ui.printTasks(foundTasks);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            System.out.println("Please enter keyword of task to search!");
+        }
```
Good catch!

joshuayapwj98 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * Types of valid commands
+     */
+    public enum CommandType {
+        LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, REMOVE, FIND, BYE, NULL
+    }
+
```
Like the use of enum here!

joshuayapwj98 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+                int isComplete = currTask.isComplete() ? 1 : 0;
+                String saveText;
+
+                if (currTask instanceof ToDo) {
+                    saveText = "T | " + isComplete + " | " + description;
+                } else if (currTask instanceof Deadline) {
+                    saveText = "D | " + isComplete + " | " + description + " | " + ((Deadline) currTask).getDeadlineBy();
+                } else if (currTask instanceof Event) {
+                    saveText = "E | " + isComplete + " | " + description + " | " + ((Event) currTask).getEventBy();
+                } else {
+                    fw.close();
```
Very neat on this part!
Similar to before, could this be transferred to the ui.java class?

Still, good work!

joshuayapwj98 reviewed (commented):
Really love the custom portion of the bot!!

Just hear for neat picking :D

whoisjunhong committed:
Fix code style and ui abstraction

whoisjunhong committed:
Merge branch 'add-gradle-support'

whoisjunhong committed:
Fix checkstyle

gigituang commented on src/main/java/duke/commands/UnmarkCommand.java:
```diff
@@ -0,0 +1,26 @@
+            System.out.println("OK, I've unmarked this task as incomplete:\n" + task);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            System.out.println("Please enter a number of the item in the list you wish to unmark!");
+        } catch (IndexOutOfBoundsException | NumberFormatException e) {
+            System.out.println("Please enter index within range 1 to " + taskList.getTasks().size());
+        }
```
I love how you emphasized the specific range instead of just stating that its out of range:)

gigituang commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -73,26 +73,26 @@ public static CommandType getCommandType(String userInput) {
+        case "bye":
+            return CommandType.BYE;
+        default:
+            return CommandType.NULL;
```
Love how you used String.split() methods to manipulate the description strings as the code sections for todo, deadline and event can be quite long

gigituang commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -36,9 +36,13 @@ public static void printNoTaskFound() {
+        System.out.println("☹ OOPS!!! command is invalid, please try again!");
     }
 
+    public static void printFindResultHeader(String keyword) {
+        System.out.println("Here are the matching tasks in your list containing "
+                + "\"" + keyword + "\"" + ":");
+    }
```
I like the structure of your codes, but it can be too lengthy when writing strings. 
I see that most of your methods prints our similar strings. Probably can try to use getMessage() method to shorten and make your strings a little more consistent. Great work overall!!

gigituang reviewed (commented):
Overall, I found your code easy to read except for the println parts where it could really lengthy. I love the naming conventions for your methods too! Keep up the good work!!

whoisjunhong committed:
Level-10: GUI

whoisjunhong committed:
Fix checkstyle

whoisjunhong committed:
Add java assertions

whoisjunhong committed:
Fix CodeQuality

whoisjunhong committed:
Add streams

whoisjunhong committed:
Merge pull request #1 from whoisjunhong/branch-A-Assertions

Add java assertions

whoisjunhong committed:
Merge pull request #2 from whoisjunhong/branch-A-CodeQuality

Fix CodeQuality

whoisjunhong committed:
Merge pull request #3 from whoisjunhong/branch-A-Streams

Add streams

whoisjunhong committed:
Merge pull request #4 from whoisjunhong/Merging-PRs

Merging PRs

whoisjunhong committed:
Add gradle.yml for github actions

whoisjunhong committed:
Add reminder function for Duke

whoisjunhong committed:
Merge pull request #5 from whoisjunhong/branch-A-Reminder

Add reminder function for Duke

whoisjunhong committed:
Fix assert blocking messages

whoisjunhong committed:
User is confused on the commands to use for the application

User will not be able to use the application as intended and deter its
user experience

HelpCommand class is added to ensure that the user is able to see the
list of commands to be use in the application. Adjust welcome message to
ensure that the help command option is seen by the user

This ensures that the user's first touch point with the application
would be able to see that there is a help command as an option to use.
This would make the user experience better and use the application as
intended by the development team.

whoisjunhong committed:
Fix UI to make it more realistic

whoisjunhong committed:
Add Ui.png

whoisjunhong committed:
User does not have a proper user guide to follow

User will get confused on how to use the application and does not have a
avenue to refer to the functionalities of the application

Wrote a extensive user guide for the user to refer to

Followed user guide conventions so that it is standardised

Refer to the full user guide here: https://whoisjunhong.github.io/ip/

whoisjunhong committed:
Fix DeleteCommand class and HelpCommand UI

whoisjunhong committed:
Task not saved properly and no dynamic file location

User's task would not be saved if they close the application abruptly,
only way was using command bye. File location to be saved is not
dynamic, user might not have data/ folder in their system. This will
cause the retriving and saving of files function to break.

All commands that require to be saved to text file will write to file
after every action. Dynamic folder and file location has been added to
prevent issues with retriving and saving of files.

This is to ensure that the functionality does not break no matter which
system the application is hosted on.

whoisjunhong committed:
Fix checkstlye

whoisjunhong committed:
Change DukeTest

whoisjunhong committed:
A-JAR-released

whoisjunhong committed:
Merge pull request #7 from whoisjunhong/branch-A-JAR-released

A-JAR-released

whoisjunhong committed:
Fix issues raised by @nus-se-bot

whoisjunhong committed:
Change user guide based on tester's comments
